server/src/Ogmios/Parser.hs

```diff
@@ -0,0 +1,146 @@
+parseResult =
+  key "result" $
+    ProtocolParameters
+      <$> parseVersion
+      <*> parseRational "decentralizationParameter"
+      -- TODO : How to parse the value `neutral` from ogmios?
+      <*> pure Nothing -- key "extraEntropy" (perhaps (makePraosNonce . BLU.fromString <$>asString))
+      <*> parseNatural "maxBlockHeaderSize"
+      <*> parseNatural "maxBlockBodySize"
+      <*> parseNatural "maxTxSize"
+      <*> parseNatural "minFeeConstant" -- I think minFeeConstant and minFeeCoefficient are swapped here
+      -- but this is consistent with the current config file.
+      <*> parseNatural "minFeeCoefficient"
+      <*> pure Nothing
+      <*> parseLovelace "stakeKeyDeposit"
+      <*> parseLovelace "poolDeposit"
+      <*> parseLovelace "minPoolCost"
+      <*> key "poolRetirementEpochBound" fromAesonParser
+      <*> parseNatural "desiredNumberOfPools"
+      <*> parseRational "poolInfluence"
+      <*> parseRational "monetaryExpansion"
+      <*> parseRational "treasuryExpansion"
+      <*> perhaps (parseLovelace "coinsPerUtxoWord")
+      <*> (key "costModels" parseCostModels <|> pure Map.empty)
+      <*> key "prices" parseExecutionPrices
+      <*> key "maxExecutionUnitsPerTransaction" parseExecutionUnits
+      <*> key "maxExecutionUnitsPerBlock" parseExecutionUnits
+      <*> perhaps (parseNatural "maxValueSize")
+      <*> perhaps (parseNatural "collateralPercentage")
+      <*> perhaps (parseNatural "maxCollateralInputs")
+
+decodeProtocolParameters :: ByteString -> Either [Text.Text] ProtocolParameters
+decodeProtocolParameters response =
+  let value :: Maybe Aeson.Value
+      value = Aeson.decode response
+   in case parseValue parseResult <$> value of
+        Just (Right params) -> Right params
+        Just (Left e) -> Left $ displayError id e
+        _ -> Left ["Fail at converting ogmios response to cardano format"]
+
+--      wrapped :: Maybe ProtocolParametersWrapper
+--      wrapped =
+--        encoded
+--          >>= Aeson.decode @ProtocolParametersWrapper
+--   in unwrapParams <$> wrapped
+
+type LocalParser a = Parsec.Parsec String () a
+
+nonZeroDigit :: LocalParser Char
+nonZeroDigit = Parsec.Char.satisfy (\c -> (c /= '0') && isDigit c)
+
+nonZeroInteger :: LocalParser Integer
+nonZeroInteger = do
+  headDigit <- nonZeroDigit
+  remains <- Parsec.many Parsec.Char.digit
+  return $ read (headDigit : remains)
+
+zeroInteger :: LocalParser Integer
+zeroInteger = read <$> many1 (Parsec.Char.satisfy (== '0'))
+
+haskellInteger :: LocalParser Integer
+haskellInteger = nonZeroInteger Parsec.<|> zeroInteger
+
+rational :: LocalParser Rational
+rational =
+  do
+    Parsec.Char.spaces
+    numerator <- haskellInteger
+    Parsec.Char.spaces >> Parsec.Char.char '/' >> Parsec.Char.spaces
+    denominator <- haskellInteger
+    return $ numerator % denominator
+
+rationalParser :: String -> Either Text.Text Rational
+rationalParser s =
+  case Parsec.runParser rational () "ogmios.json" s of
```

Comment:
`either` would work here instead of an explicit case expression

Reply:
We want to return an `Either` here to propagate the error to `Aeson.BetterError`, so, maybe you were thinking of `first` from the `Bifunctor` instance of `Either`?